File: python/ml.py
import numpy as np
import pandas as pd
from matplotlib import pyplot as plt
from icecream import ic
import json



# is full pipeline from backup file same as main2?
# I can add pipeline() if necessary.
def main2():
    # Feedforward Neural Network (FNN)     

    data = pd.read_csv('../digit-recognizer/train.csv')
    data = np.array(data)
    m, n = data.shape
    print("data.shape:", data.shape)

    np.random.shuffle(data)

    # dev set
    data_dev = data[0:1000].T
    Y_dev = data_dev[0]
    X_dev = data_dev[1:n]
    X_dev = X_dev / 255.

    print(Y_dev[:5])
    print(X_dev[0][:5])

    # training set 
    data_train = data[1000:m].T
    Y_train = data_train[0]
    X_train = data_train[1:n]
    X_train = X_train / 255.
    _,m_train = X_train.shape

    print("Y_train:", Y_train)
    print("m_train:", m_train)

    W1, b1, W2, b2 = gradient_descent(X_train, Y_train, 0.10, 500, m)

    test_prediction(0, W1, b1, W2, b2, X_train, Y_train)
    test_prediction(1, W1, b1, W2, b2, X_train, Y_train)
    test_prediction(2, W1, b1, W2, b2, X_train, Y_train)
    test_prediction(3, W1, b1, W2, b2, X_train, Y_train)

    print("Y_dev:", Y_dev[:5])           # Print first 5 elements of Y_dev for brevity
    print("X_dev sample:", X_dev[0][:5]) # Print first 5 elements of the first row of X_dev for brevity

    dev_predictions = make_predictions(X_dev, W1, b1, W2, b2)
    print(get_accuracy(dev_predictions, Y_dev))

    output_json(W1, b1, W2, b2)



def gradient_descent(X, Y, alpha, iterations, m):
    W1, b1, W2, b2 = init_params()
    for i in range(iterations):
        Z1, A1, Z2, A2 = forward_prop(W1, b1, W2, b2, X)
        dW1, db1, dW2, db2 = backward_prop(Z1, A1, Z2, A2, W1, W2, X, Y, m)
        W1, b1, W2, b2 = update_params(W1, b1, W2, b2, dW1, db1, dW2, db2, alpha)
        if i % 10 == 0:
            predictions = get_predictions(A2)
            print(get_accuracy(predictions, Y))
    return W1, b1, W2, b2

def init_params():
    W1 = np.random.rand(10, 784) - 0.5
    b1 = np.random.rand(10, 1) - 0.5
    W2 = np.random.rand(10, 10) - 0.5
    b2 = np.random.rand(10, 1) - 0.5
    print("\ninit_params():")
    print(f"type(W1):{type(W1)}, type(b1): {type(b1)}, type(W2): {type(W2)}, type(b2): {type(b2)}")
    return W1, b1, W2, b2

def forward_prop(W1, b1, W2, b2, X):
    Z1 = W1.dot(X) + b1
    A1 = ReLU(Z1)
    Z2 = W2.dot(A1) + b2
    A2 = softmax(Z2)
    return Z1, A1, Z2, A2

def backward_prop(Z1, A1, Z2, A2, W1, W2, X, Y, m):
    one_hot_Y = one_hot(Y)
    dZ2 = A2 - one_hot_Y
    dW2 = 1 / m * dZ2.dot(A1.T)
    db2 = 1 / m * np.sum(dZ2, axis=1).reshape(-1, 1)
    dZ1 = W2.T.dot(dZ2) * ReLU_deriv(Z1)
    dW1 = 1 / m * dZ1.dot(X.T)
    db1 = 1 / m * np.sum(dZ1, axis=1).reshape(-1, 1)
    return dW1, db1, dW2, db2

def update_params(W1, b1, W2, b2, dW1, db1, dW2, db2, alpha):
    W1 = W1 - alpha * dW1
    b1 = b1 - alpha * db1    
    W2 = W2 - alpha * dW2  
    b2 = b2 - alpha * db2    
    return W1, b1, W2, b2

def ReLU(Z):
    return np.maximum(Z, 0)

def softmax(Z):
    A = np.exp(Z) / sum(np.exp(Z))
    return A

def ReLU_deriv(Z):
    return Z > 0

def one_hot(Y):
    one_hot_Y = np.zeros((Y.size, Y.max() + 1))
    one_hot_Y[np.arange(Y.size), Y] = 1
    one_hot_Y = one_hot_Y.T
    return one_hot_Y

def get_predictions(A2):
    return np.argmax(A2, 0)

def get_accuracy(predictions, Y):
    print(predictions, Y)
    return np.sum(predictions == Y) / Y.size

def make_predictions(X, W1, b1, W2, b2):
    _, _, _, A2 = forward_prop(W1, b1, W2, b2, X)
    predictions = get_predictions(A2)
    return predictions

def test_prediction(index, W1, b1, W2, b2, X_train, Y_train):
    current_image = X_train[:, index, None]
    prediction = make_predictions(X_train[:, index, None], W1, b1, W2, b2)
    label = Y_train[index]
    print("Prediction: ", prediction)
    print("Label: ", label)
    
    current_image = current_image.reshape((28, 28)) * 255
    plt.gray()
    plt.imshow(current_image, interpolation='nearest')
    plt.show()


def output_json(W1, b1, W2, b2):

    weights_biases = {
        "W1": W1.tolist(),
        "b1": b1.tolist(),
        "W2": W2.tolist(),
        "b2": b2.tolist(),
    }

    json_string = json.dumps(weights_biases, indent=4)

    filename = '../weights_and_biases.json'
    with open(filename, 'w') as file:
        file.write(json_string)

File: golang/ml.go
package main

// train in python, paralellize is python?, save to json, load in go, run prediction in go, parallelize in go?

import (
	"encoding/csv"
	"fmt"
	"log"
	"math/rand"
	"os"
	"time"

	"net/http"
	_ "net/http/pprof"
)

func pipeline() { // rename to main()

	go func() {
		log.Println(http.ListenAndServe("localhost:6060", nil))
	}()

	fmt.Println("Running neural_network_from_scratch_in_go")

	// Load training data
	file, err := os.Open("../digit-recognizer/train.csv")
	if err != nil {
		fmt.Println("Error:", err)
		return
	}
	defer file.Close()

	reader := csv.NewReader(file)
	records, err := reader.ReadAll()
	if err != nil {
		fmt.Println("Error:", err)
		return
	}

	// Remove the header row
	records = records[1:]

	// Determine the dimensions of records
	numRows := len(records)
	numCols := 0
	if numRows > 0 {
		numCols = len(records[0]) // Assuming uniformity in the slice
	}
	fmt.Printf("Shape of records: %d rows x %d columns\n", numRows, numCols)

	// Convert [][]string to [][]float64
	floatRecords, err := convertToFloat64(records)
	if err != nil {
		fmt.Println("Error converting records:", err)
		return
	}

	_, _ = getDimensions2DFloat(floatRecords)

	// Seed the random number generator to ensure different results each run
	rand.Seed(time.Now().UnixNano())

	// Randomly shuffle the rows of the 2D slice
	rand.Shuffle(len(floatRecords), func(i, j int) {
		floatRecords[i], floatRecords[j] = floatRecords[j], floatRecords[i]
	})

	// // Print the converted records
	// fmt.Println(floatRecords)

	dataSliced := floatRecords[:1000]
	dataTransposed := transpose(dataSliced)
	dataTransposedInt := transposeInt(float64ToInt(dataSliced))
	Y_dev := dataTransposedInt[0]
	X_dev := dataTransposed[1:numCols]

	// Dividing each element of X_dev by 255
	for i := range X_dev {
		for j := range X_dev[i] {
			X_dev[i][j] /= 255.0
		}
	}

	// Printing Y_dev and a portion of X_dev to verify
	fmt.Println("Y_dev:", Y_dev[:5])           // Print first 5 elements of Y_dev for brevity
	fmt.Println("X_dev sample:", X_dev[0][:5]) // Print first 5 elements of the first row of X_dev for brevity

	// Slicing data[1000:num_rows] and Transposing
	dataSliced = floatRecords[1000:numRows]
	dataTrain := transpose(dataSliced)
	dataTrainInt := transposeInt(float64ToInt((dataSliced)))
	Y_train := dataTrainInt[0]
	X_train := dataTrain[1:numCols] // Assuming num_cols is within bounds

	// Dividing each element of X_train by 255
	for i := range X_train {
		for j := range X_train[i] {
			X_train[i][j] /= 255.0
		}
	}

	// Determining the shape of X_train
	mNum := len(X_train[0]) // Number of columns in X_train

	// Printing Y_train
	fmt.Println("Y_train:", Y_train)

	// Printing the shape of X_train
	fmt.Printf("mNum: %d", mNum)

	W1, b1, W2, b2 := gradientDescent(X_train, Y_train, 0.10, 500, mNum)

	testPrediction(0, W1, b1, W2, b2, X_train, Y_train)
	testPrediction(1, W1, b1, W2, b2, X_train, Y_train)
	testPrediction(2, W1, b1, W2, b2, X_train, Y_train)
	testPrediction(3, W1, b1, W2, b2, X_train, Y_train)

	devPredictions := makePredictions(X_dev, W1, b1, W2, b2)
	print(getAccuracy(devPredictions, Y_dev))

}

func runPredictions(data Data, floatRecords [][]float64, numCols int) {
	// Randomly shuffle the rows of the 2D slice
	rand.Shuffle(len(floatRecords), func(i, j int) {
		floatRecords[i], floatRecords[j] = floatRecords[j], floatRecords[i]
	})

	// // Print the converted records
	// fmt.Println(floatRecords)

	dataSliced := floatRecords[:1000]
	dataTransposed := transpose(dataSliced)
	dataTransposedInt := transposeInt(float64ToInt(dataSliced))
	Y_dev := dataTransposedInt[0]
	X_dev := dataTransposed[1:numCols]

	// Dividing each element of X_dev by 255
	for i := range X_dev {
		for j := range X_dev[i] {
			X_dev[i][j] /= 255.0
		}
	}

	// Printing Y_dev and a portion of X_dev to verify
	fmt.Println("Y_dev:", Y_dev[:5])           // Print first 5 elements of Y_dev for brevity
	fmt.Println("X_dev sample:", X_dev[0][:5]) // Print first 5 elements of the first row of X_dev for brevity

	devPredictions := makePredictions(X_dev, data.W1, data.B1, data.W2, data.B2)
	acc := getAccuracy(devPredictions, Y_dev)
	fmt.Printf("Accuracy: %.2f\n", acc)
}

func makePredictions(X, W1, b1, W2, b2 [][]float64) []int {
	_, _, _, A2 := forwardProp(W1, b1, W2, b2, X)
	predictions := getPredictions(A2)
	return predictions
}

// getAccuracy calculates the accuracy as the proportion of correct predictions.
// It assumes predictions and Y are slices of the same length.
func getAccuracy(predictions, Y []int) float64 {
	if len(predictions) == 0 || len(predictions) != len(Y) {
		return 0.0 // Return 0 if slices are empty or not of the same length.
	}

	correctCount := 0
	for i, prediction := range predictions {
		if prediction == Y[i] {
			correctCount++
		}
	}

	accuracy := float64(correctCount) / float64(len(Y))
	return accuracy
}

// getPredictions returns the indices of the maximum values in each column of a 2D slice.
func getPredictions(A2 [][]float64) []int {
	if len(A2) == 0 || len(A2[0]) == 0 {
		return nil // Return nil if A2 is empty or has empty sub-slices.
	}

	rows := len(A2)
	cols := len(A2[0])
	maxIndices := make([]int, cols)

	for j := 0; j < cols; j++ {
		maxIndex := 0
		maxValue := A2[0][j]
		for i := 1; i < rows; i++ {
			if A2[i][j] > maxValue {
				maxValue = A2[i][j]
				maxIndex = i
			}
		}
		maxIndices[j] = maxIndex
	}

	return maxIndices
}

// forwardProp performs forward propagation through a 2-layer neural network.
func forwardProp(W1, b1, W2, b2, X [][]float64) (
	Z1, A1, Z2, A2 [][]float64,
) {
	intermed := matrixDot(W1, X)
	Z1 = broadcastAdd(intermed, b1)
	A1 = ReLU(Z1)
	Z2 = broadcastAdd(matrixDot(W2, A1), b2)
	A2 = softmax(Z2)
	return Z1, A1, Z2, A2
}

// initParams initializes and returns matrices W1, b1, W2, b2
func initParams() ([][]float64, [][]float64, [][]float64, [][]float64) {
	rand.Seed(time.Now().UnixNano())

	W1 := make([][]float64, 10)
	b1 := make([][]float64, 10)
	for i := range W1 {
		W1[i] = make([]float64, 784)
		b1[i] = make([]float64, 1)
		for j := range W1[i] {
			W1[i][j] = rand.Float64() - 0.5
		}
		b1[i][0] = rand.Float64() - 0.5
	}

	W2 := make([][]float64, 10)
	b2 := make([][]float64, 10)
	for i := range W2 {
		W2[i] = make([]float64, 10)
		b2[i] = make([]float64, 1)
		for j := range W2[i] {
			W2[i][j] = rand.Float64() - 0.5
		}
		b2[i][0] = rand.Float64() - 0.5
	}

	return W1, b1, W2, b2
}

// backwardProp performs backward propagation through a 2-layer neural network.
func backwardProp(Z1, A1, A2, W2, X [][]float64, Y []int, mNum int) (
	dW1, db1, dW2, db2 [][]float64,
) {
	oneHotY := oneHot2(Y, len(A2))
	fmt.Println("Passed 0")

	dZ2 := matrixSubtract(A2, oneHotY)
	dW2 = matrixDot(transpose(dZ2), A1)
	for i := range dW2 {
		for j := range dW2[i] {
			dW2[i][j] /= float64(mNum)
		}
	}
	fmt.Println("Passed 1")
	db2 = make([][]float64, 1)
	db2[0] = matrixMean(dZ2, mNum)

	fmt.Println("Passed 1.5")

	dZ1 := matrixDot(W2, dZ2)
	reluDerivZ1 := ReLUderiv(Z1)
	for i := range dZ1 {
		for j := range dZ1[i] {
			dZ1[i][j] *= reluDerivZ1[i][j]
		}
	}
	fmt.Println("Passed 2")

	dW1 = matrixDot(transpose(dZ1), X)
	fmt.Println("Passed 2.5")
	for i := range dW1 {
		for j := range dW1[i] {
			dW1[i][j] /= float64(mNum)
		}
	}
	fmt.Println("Passed 3")

	db1Sum := matrixMean(dZ1, mNum)
	db1 = make([][]float64, len(db1Sum))
	for i, val := range db1Sum {
		db1[i] = []float64{val}
	}
	fmt.Println("Passed 4")

	return dW1, db1, dW2, db2
}

// updateParams updates the parameters by subtracting the gradients scaled by alpha.
// Assumes W1, b1, W2, b2, db1, dW2, db2 are all slices of slices of float64.
// alpha is a float64 scalar.
func updateParams(W1, b1, W2, b2, db1, dW2, db2 [][]float64, alpha float64) ([][]float64, [][]float64, [][]float64, [][]float64) {
	// Update b1
	for i := range b1 {
		for j := range b1[i] {
			b1[i][j] = b1[i][j] - alpha*db1[i][j]
		}
	}

	// Update W2
	for i := range W2 {
		for j := range W2[i] {
			W2[i][j] = W2[i][j] - alpha*dW2[i][j]
		}
	}

	// Update b2
	for i := range b2 {
		for j := range b2[i] {
			b2[i][j] = b2[i][j] - alpha*db2[i][j]
		}
	}

	return W1, b1, W2, b2
}

func gradientDescent(
	X [][]float64,
	Y []int,
	alpha float64,
	iterations int,
	mNum int,
) (
	W1, b1, W2, b2 [][]float64,
) {
	W1, b1, W2, b2 = initParams()
	for i := 0; i < iterations; i++ {
		fmt.Println("i:", i)
		Z1, A1, _, A2 := forwardProp(W1, b1, W2, b2, X)
		fmt.Println("passed forwardProp")
		_, db1, dW2, db2 := backwardProp(Z1, A1, A2, W2, X, Y, mNum)
		fmt.Println("passed backwardProp")
		W1, b1, W2, b2 = updateParams(W1, b1, W2, b2, db1, dW2, db2, alpha)
		fmt.Println("passed updateParams")
		if i%10 == 0 {
			fmt.Println("Iteration:", i)
			predictions := getPredictions(A2)
			fmt.Println(getAccuracy(predictions, Y))
		}
	}
	return W1, b1, W2, b2
}

func testPrediction(
	index int,
	W1, b1, W2, b2, X_train [][]float64,
	Y_train []int,
) {
	current_image := extractColumn(X_train, index)
	prediction := makePredictions(current_image, W1, b1, W2, b2)
	label := Y_train[index]
	fmt.Println("Prediction: ", prediction)
	fmt.Println("Label: ", label)
}

// Refactor redundancy
func loadDevData(floatRecords [][]float64, numCols int) ([][]float64, []int) {
	// Randomly shuffle the rows of the 2D slice
	rand.Shuffle(len(floatRecords), func(i, j int) {
		floatRecords[i], floatRecords[j] = floatRecords[j], floatRecords[i]
	})

	// // Print the converted records
	// fmt.Println(floatRecords)

	dataSliced := floatRecords[:1000]
	dataTransposed := transpose(dataSliced)
	dataTransposedInt := transposeInt(float64ToInt(dataSliced))
	Y_dev := dataTransposedInt[0]
	X_dev := dataTransposed[1:numCols]

	// Dividing each element of X_dev by 255
	for i := range X_dev {
		for j := range X_dev[i] {
			X_dev[i][j] /= 255.0
		}
	}

	// Printing Y_dev and a portion of X_dev to verify
	fmt.Println("Y_dev:", Y_dev[:5])           // Print first 5 elements of Y_dev for brevity
	fmt.Println("X_dev sample:", X_dev[0][:5]) // Print first 5 elements of the first row of X_dev for brevity

	return X_dev, Y_dev
}
